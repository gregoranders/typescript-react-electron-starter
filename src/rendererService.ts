import { ipcMain, ipcRenderer } from "electron";

import { IIPCMessage, IPCMessageType, mainChannel, renderChannel } from "./ipc";

export class RendererService {

  public ping(): void {
    this.send("ping", [1, 2, 3]);
  }

  public register(): void {
    ipcRenderer.on(renderChannel, (event: Electron.IpcRendererEvent, ...args: any[]): void => {
      if (args && args.length === 1) {
        const ipcMessage: IIPCMessage<any> = args[0];
        this.handleRenderChannel(event, ipcMessage);
      } else {
        console.log(event, args);
      }
    });
  }

  private handleRenderChannel<T>(event: Electron.IpcRendererEvent, message: IIPCMessage<T>): void {
    switch (message.type) {
      case "ping":
        console.log(message);
        this.send("pong", message.data);
        break;
      case "pong":
        console.log(message);
        break;
      default:
        console.log(message);
    }
  }

  private send<T>(type: IPCMessageType, data: T): void {
    ipcRenderer.send(mainChannel, {
      data,
      timestamp: new Date(),
      type,
    } as IIPCMessage<T>);
  }
}
