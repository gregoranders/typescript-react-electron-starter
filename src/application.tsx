import * as React from "react";

import { RendererService } from "./rendererService";

import { Button } from "./button";

interface IApplicationProperties {
  service: RendererService;
}

export class Application extends React.Component<IApplicationProperties> {

  public render(): React.ReactNode {
    return (<div className="section container">
      <h1>Application loaded</h1>
      <Button label="Toggle" onClick={(): void => { this.handleButtonClick(); }}/>
    </div>);
  }

  private handleButtonClick(): void {
    this.props.service.ping();
  }
}
