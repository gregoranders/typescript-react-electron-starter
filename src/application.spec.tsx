import { mount, ReactWrapper } from "enzyme";
import * as React from "react";
import * as renderer from "react-test-renderer";

import * as TestSubject from "./application";

import { RendererService } from "./rendererService";

describe("Application", (): void => {

  const pingMock: jest.Mock<any, any[]> = jest.fn();
  const RendererServiceMock: jest.Mock<RendererService, any[]>
    = jest.fn().mockImplementation((props: TestSubject.IApplicationProperties, context?: any) => {
      return {
        ping: pingMock,
      };
  });

  const service: RendererService = new RendererServiceMock();

  beforeEach((): void => {
    RendererServiceMock.mockClear();
  });
  it("export exists", (): void => {
    expect(TestSubject.Application).toBeDefined();
  });

  it("snapshot", (): void => {
    const snapshot: renderer.ReactTestRenderer = renderer.create(<TestSubject.Application service={ service } />);
    snapshot.toJSON();
    expect(snapshot).toMatchSnapshot();
  });

  it("header", (): void => {
    const testSubject: ReactWrapper<TestSubject.IApplicationProperties, {}, TestSubject.Application>
      = mount(<TestSubject.Application service={ service } />);

    expect(testSubject.find("h1").text()).toBe("Application loaded");
  });

  it("button clicked", (): void => {
    const testSubject: ReactWrapper<TestSubject.IApplicationProperties, {}, TestSubject.Application>
      = mount(<TestSubject.Application service={ service } />);

    expect(pingMock).toBeCalledTimes(0);
    testSubject.find("Button").find("button").simulate("click");
    expect(pingMock).toBeCalledTimes(1);
  });

});
