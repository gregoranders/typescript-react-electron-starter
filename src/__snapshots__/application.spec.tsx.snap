// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Application snapshot 1`] = `
<div
  className="section container"
>
  <h1>
    Application loaded
  </h1>
  <div>
    <h2>
      1
    </h2>
    <button
      className="btn btn-primary"
      onClick={[Function]}
    >
      Toggle
    </button>
  </div>
</div>
`;
